packages/contracts/contracts/ETHVirtualAppAgreement.sol

```diff
@@ -40,50 +33,31 @@ contract ETHVirtualAppAgreement {
       "agreement lockup time has not elapsed"
     );
 
-    Transfer.Transaction memory resolution = agreement.registry
+    bytes memory resolution = agreement.registry
       .getResolution(agreement.appIdentityHash);
 
-    require(
-      agreement.terms.assetType == resolution.assetType,
-      "returned incompatible resolution based on bad asset type"
-    );
-
-    require(
-      agreement.terms.token == resolution.token,
-      "returned incompatible resolution based on bad token value"
-    );
-
-    require(
-      agreement.capitalProvided > resolution.value[0],
-      "returned incompatible resolution"
-    );
+    uint256 resolutionAsValue = abi.decode(resolution, (uint256));
 
     require(
       !agreement.nonceRegistry.isFinalizedOrHasNeverBeenSetBefore(agreement.uninstallKey, 1),
       "Virtual app agreement has been uninstalled"
     );
 
-    uint256[] memory amount = new uint256[](2);
-
-    amount[0] = resolution.value[0];
-    amount[1] = agreement.capitalProvided - amount[0];
+    if (resolutionAsValue == 0) {
+      agreement.beneficiaries[0].send(agreement.capitalProvided);
+      return;
+    } else if (resolutionAsValue == 1) {
+      agreement.beneficiaries[1].send(agreement.capitalProvided);
+      return;
+    }
 
-    bytes[] memory data = new bytes[](2);
+    /* SPLIT_AND_SEND_TO_BOTH_ADDRS or default cases */
 
-    address[] memory beneficiaries = new address[](2);
-
-    beneficiaries[0] = agreement.beneficiaries[0];
-    beneficiaries[1] = agreement.beneficiaries[1];
-
-    Transfer.Transaction memory ret = Transfer.Transaction(
-      agreement.terms.assetType,
-      agreement.terms.token,
-      beneficiaries,
-      amount,
-      data
-    );
+    agreement.beneficiaries[0].transfer(agreement.capitalProvided / 2);
+    agreement.beneficiaries[1].transfer(
+      agreement.capitalProvided - agreement.capitalProvided / 2);
 
-    ret.execute();
+    return;
```

Comment:
So you are choosing to basically embed the interpreter's execution function inside of this special case contract for virtual apps. Does this hurt our ability to fully generalize virtual apps with `StateChannelTransaction` in the future?

Reply:
The exact same set of execution functions cannot be shared, because of the fact that for e.g. NFTs cannot be routed across a channel network